src/framework/components/anim/controller.js

```diff
@@ -0,0 +1,438 @@
+            this.timeInStateBefore = timeInStateBefore;
+        },
+
+        _transitionToState: function(newStateName) {
+            if (newStateName === this.activeStateName) {
+                return;
+            }
+
+            if (!this._getState(newStateName)) {
+                return;
+            }
+
+            var transition = this._findTransition(this.activeStateName, newStateName);
+            if (!transition) {
+                this.animEvaluator.removeClips();
+                transition = new AnimTransition(this, this.activeStateName, newStateName, 0, 0);
+            }
+            this._updateStateFromTransition(transition);
+        },
+
+        _transitionToNextState: function() {
+            var transition = this._findTransition();
+            if (!transition) {
+                return;
+            }
+            if (transition.to === 'End')
+            {
+                this._setActiveState('Start');
+                transition = this._findTransition();
+            }
+            this._updateStateFromTransition(transition);
+        },
+
+        linkAnimationToState: function(stateName, animTrack) {
+            var state = this._getState(stateName);
+            if (!state) {
+                console.error('Linking animation asset to animation state that does not exist');
+                return;
+            }
+
+            var animation = {
+                name: stateName + '.' + animTrack.name,
+                animTrack: animTrack,
+                weight: 1.0
+            };
+
+            // Currently the anim controller only supports single animations in a state
+            if (state.animations.length > 0) {
+                state.animations = [];
+                this.reset();
+            }
+            state.animations.push(animation);
+
+            if (!this.playing && this.activate && this.isPlayable()) {
+                this.play();
+                return;
+            }
+        },
+
+        isPlayable: function() {
+            var playable = true;
+            for (var i = 0; i < this.states.length; i++) {
+                if (!this.states[i].isPlayable()) {
+                    playable = false;
+                }
+            }
+            return playable;
+        },
+
+        play: function(stateName) {
+            if (stateName) {
+                this._transitionToState(stateName);
+            }
+            this.playing = true;
+        },
+
+        reset: function() {
+            this.previousStateName = null;
+            this.activeStateName = 'Start';
+            this.playing = false;
+            
+            this.currTransitionTime = 1.0;
+            this.totalTransitionTime = 1.0;
+            this.isTransitioning = false;
+
+            this.timeInState = 0;
+            this.timeInStateBefore = 0;
+
+            this.animEvaluator.removeClips();
+
+            this.parameters = JSON.parse(JSON.stringify(this.initialParameters));
```

Comment:
wha?! I'm quite sure you don't what to be doing this!

Reply:
The aim is to deep copy the parameters so I can set them back to the initial values on reset.

Reply:
Oh yeah I realize the intent :) stringify and parse though will likely be real slow. I would assume there already exists a 'deepCopy' type operation in the engine somewhere otherwise I'd suggest you add one.

Actually it's not as slow as I expected https://jsben.ch/bWfk9, but it looks like you could just use Object.assign?

Reply:
Nice website. Shallow copy will be fine here as the parameters aren't storing any objects atm. I'll switch this to use Object.assign.